### ISSUE TITLE
[Bug] NetPak issues / Effect is oriented incorrectly

### ISSUE THREAD

**Issue by Gamingtoday093** [SmartlyDressedGames/Unturned-3.x-Community#3686]
I've continued the development of my silly little WayFinder plugin, and I've unfortunately encountered *another* issue with effects.

Thankfully this time it actually happens somewhat consistently. The problem is that some of the effects do not spawn in their correct orientation under certain conditions.
![ProblemEffect](https://user-images.githubusercontent.com/70017362/216774984-2a71a00a-2e47-4ee2-aed7-63e5ffaa6267.png)
*(It's supposed to point straight ahead like the one behind it.)*

***It's not related to RandomizeRotation. It's turned off for this asset, and its rotation consistently spawned in the exact same direction.***

So I originally started writing this report about 2-3 hours ago. which at that time was really just a report about this weird behavior with no real insight into why. I did some more testing, and a lot of new things became apparent to me.

Such as that `Vector3.ToString()` shortens the position floats to the first decimal even if it has a value smaller than `0.1`. The problem effect described above is highlighted in red.
![Hidden](https://user-images.githubusercontent.com/70017362/216779975-1f16b183-d11c-429a-89cf-1f4c8950140d.png)
The actual components of the Vector3 are:
![Showing](https://user-images.githubusercontent.com/70017362/216780052-ede6887f-9561-48cb-95bc-56ac9dd839ac.png)
which surprised me since the Y component really should be 0.

This led me to fine comb through my code until I realized that `Direction += *Another Vector3*` added some slight errors in the Y component. Setting the Y component of the Direction vector to 0 fixed the weird orientation issue. But I really wasn't happy with that solution since now my goal was to figure out why this happened. I then, by random chance, happened to use the `Direction` vector directly instead of `Direction.normalized` as I had been using to set the direction of the effect. Which completely solved the issue?!

This got me thinking that it may be due to some weird NetPak data loss, so I continued my quest to solve this really strange issue. I set up some basic read and write Netpak classes, and lo and behold, I was right!

Ignore the red-highlighted values. I really struggled to get this working, and in the end I only managed to make the X value of the Vector3 be transported. In this case, it's the Y component of the Default (non-normalized) vector. which got transported perfectly fine (with some slight errors).
![Default](https://user-images.githubusercontent.com/70017362/216780616-86a9ad78-33bf-40e5-a891-ccc0150b2b33.png)
However, using the normalized version of the same vector returns an unexpected result.
![Normalized](https://user-images.githubusercontent.com/70017362/216781111-af7a6ec2-2867-43b4-8f32-5bc77b78842f.png)

For some reason, the Y component is set to `1` after being net transported. Super strange behaviour that I got no real idea on why is happening. So I say that this is where my journey ends, I am really not familiar with any of this net stuff or bit shifting. Still, I'd consider it a success since I've kind of found the explanation to why it happens.

Also yes. I understand that there will be some data loss related to transporting things over the internet, but this large of a discrepancy must surely be a bug or an oversight.

**Comment by SDGNelson** [comment]
Thanks, will check into it! The net pak stuff should be fairly stable: there are actually quite a few unit tests for it. The effect code is really old and should probably be updated to use quaternions. Then EffectSpawnParameters would be updated with helpers for quats, euler angles, and direction as suggested in your idea #3687

**Comment by SDGNelson** [comment]
Your guess was correct that there was a bug when reading/writing that vector!

I've now added ~20 more test cases with extremely small floats like `-2.060574E-13f`, including the specific vector from your example. 

Ran into something I haven't experienced in the last ~10 years (feels weird to write that): logging the related values prevented the bug. Calling `Debug.Log` with `intValue` and `fracValue` in `WriteClampedFloat` prevents the bug. I double-checked this in source control adding/removing changed the behaviour. Still not sure why. :confused:

Anyway it seems the problem was flooring and subtracting the value behaved unexpectedly / precision issues. ClampedFloat is only used by positions, and I had been meaning to add a parameter attribute for normal vectors, so ClampedFloat now rounds values very near zero (0.0001) and most if not all normal vectors in the base game should use the different serialization.
